File: packages/google_mobile_ads/lib/src/ump/user_messaging_codec.dart
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';
import 'package:google_mobile_ads/src/ump/consent_form.dart';

import 'consent_information.dart';
import 'consent_request_parameters.dart';

/// Codec for coding and decoding types in the ump sdk.
class UserMessagingCodec extends StandardMessageCodec {

  static const int _valueConsentInformation = 128;
  static const int _valueConsentRequestParameters = 129;
  static const int _valueConsentDebugSettings = 130;
  static const int _valueConsentForm = 131;

  @override
  void writeValue(WriteBuffer buffer, dynamic value) {
    if (value is ConsentRequestParameters) {
      buffer.putUint8(_valueConsentRequestParameters);
      writeValue(buffer, value.tagForUnderAgeOfConsent);
      writeValue(buffer, value.consentDebugSettings);
    } else if (value is ConsentDebugSettings) {
      buffer.putUint8(_valueConsentDebugSettings);
      writeValue(buffer, value.debugGeography);
    } else if (value is ConsentInformation) {
      buffer.putUint8(_valueConsentInformation);
      writeValue(buffer, value.hash);
    } else if (value is ConsentForm) {
      buffer.putUint8(_valueConsentForm);
      writeValue(buffer, value.hash);
    } else {
      super.writeValue(buffer, value);
    }
  }

  @override
  dynamic readValueOfType(dynamic type, ReadBuffer buffer) {
    switch (type) {
      case _valueConsentInformation:
        final int hashCode = readValueOfType(buffer.getUint8(), buffer);
        return ConsentInformation(hashCode);
      case _valueConsentForm:
        final int hashCode = readValueOfType(buffer.getUint8(), buffer);
        return ConsentForm(hashCode);
      default:
        return super.readValueOfType(type, buffer);
    }
  }
}

File: packages/google_mobile_ads/android/src/main/java/io/flutter/plugins/googlemobileads/usermessagingplatform/UserMessagingCodec.java
// Copyright 2022 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package io.flutter.plugins.googlemobileads.usermessagingplatform;

import android.content.Context;
import androidx.annotation.NonNull;
import com.google.android.ump.ConsentDebugSettings;
import com.google.android.ump.ConsentForm;
import com.google.android.ump.ConsentInformation;
import com.google.android.ump.ConsentRequestParameters;
import io.flutter.plugin.common.StandardMessageCodec;
import java.io.ByteArrayOutputStream;
import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class UserMessagingCodec extends StandardMessageCodec {

  private static final byte VALUE_CONSENT_INFORMATION = (byte) 128;
  private static final byte VALUE_CONSENT_REQUEST_PARAMETERS = (byte) 129;
  private static final byte VALUE_CONSENT_DEBUG_SETTINGS = (byte) 130;
  private static final byte VALUE_CONSENT_FORM = (byte) 131;

  private final Context context;

  private final Map<Integer, ConsentForm> consentFormMap;
  private final Map<Integer, ConsentInformation> consentInformationMap;

  UserMessagingCodec(@NonNull Context context) {
    this.context = context;
    consentFormMap = new HashMap<>();
    consentInformationMap = new HashMap<>();
  }

  @Override
  protected void writeValue(ByteArrayOutputStream stream, Object value) {
    if (value instanceof ConsentInformation) {
      stream.write(VALUE_CONSENT_INFORMATION);
      consentInformationMap.put(value.hashCode(), (ConsentInformation) value);
      writeValue(stream, value.hashCode());
    } else if (value instanceof ConsentForm) {
      stream.write(VALUE_CONSENT_FORM);
      consentFormMap.put(value.hashCode(), (ConsentForm) value);
      writeValue(stream, value.hashCode());
    } else {
      super.writeValue(stream, value);
    }
  }

  @Override
  protected Object readValueOfType(byte type, ByteBuffer buffer) {
    switch (type) {
      case VALUE_CONSENT_INFORMATION:
        {
          Integer hash = (Integer) readValueOfType(buffer.get(), buffer);
          return consentInformationMap.get(hash);
        }
      case VALUE_CONSENT_REQUEST_PARAMETERS:
        {
          Boolean tfuac = (Boolean) readValueOfType(buffer.get(), buffer);
          ConsentDebugSettings debugSettings =
              (ConsentDebugSettings) readValueOfType(buffer.get(), buffer);
          ConsentRequestParameters.Builder builder = new ConsentRequestParameters.Builder();

          if (tfuac != null) {
            builder.setTagForUnderAgeOfConsent(tfuac);
          }
          if (debugSettings != null) {
            builder.setConsentDebugSettings(debugSettings);
          }
          return builder.build();
        }
      case VALUE_CONSENT_DEBUG_SETTINGS:
        {
          Integer settings = (Integer) readValueOfType(buffer.get(), buffer);
          List<String> testIdentifiers = (List<String>) readValueOfType(buffer.get(), buffer);
          ConsentDebugSettings.Builder builder = new ConsentDebugSettings.Builder(context);
          if (settings != null) {
            builder.setDebugGeography(settings);
          }
          if (testIdentifiers != null) {
            for (String testIdentifier : testIdentifiers) {
              builder.addTestDeviceHashedId(testIdentifier);
            }
          }
          return builder.build();
        }
      case VALUE_CONSENT_FORM:
        {
          Integer hash = (Integer) readValueOfType(buffer.get(), buffer);
          return consentFormMap.get(hash);
        }
      default:
        return super.readValueOfType(type, buffer);
    }
  }
}
